### PR TITLE
Port over upwind flux for curved scalar waves as a BoundaryCorrection

### PR DESCRIPTION
**Pull request by prayush**
## Proposed changes

This PR ports over upwind penalty flux for the curved scalar wave evolution system to the `BoundaryCorrection` framework. In the process, it modifies the time derivatives' calculator action struct to furnish additional quantities needed by the upwind flux.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by nikwit** [comment]
Have you tried running this with a test executable? I havent looked at the code too much but when I run this with the constraint preserving boundary conditions I ported on a simple sine example with Minkowski background, the whole wave explodes almost immediately everywhere on the domain. When I just copy over the UpwindPenalty from the (flat) `ScalarWave` system it works fine. This is just a very naive substitution so maybe I am doing something wrong though.

Maybe there could be a unit test which ensures that the flux reduces to the flat system for Minkowski?

**Comment by prayush** [comment]
> Have you tried running this with a test executable? I havent looked at the code too much but when I run this with the constraint preserving boundary conditions I ported on a simple sine example with Minkowski background, the whole wave explodes almost immediately everywhere on the domain. When I just copy over the UpwindPenalty from the (flat) `ScalarWave` system it works fine. This is just a very naive substitution so maybe I am doing something wrong though.
> 
> Maybe there could be a unit test which ensures that the flux reduces to the flat system for Minkowski?

Hi @nikwit - I found and fixed a major bug in the code here, the internal and external contributions were being added instead of subtracted. I wrote a test comparing this flux class with the flat `ScalarWave`'s upwind flux in flat space, and that helped me discover this bug. After fixing it, that test now passes. This code is now ready for review.

**Comment by prayush** [comment]
> > Have you tried running this with a test executable? I havent looked at the code too much but when I run this with the constraint preserving boundary conditions I ported on a simple sine example with Minkowski background, the whole wave explodes almost immediately everywhere on the domain. When I just copy over the UpwindPenalty from the (flat) `ScalarWave` system it works fine. This is just a very naive substitution so maybe I am doing something wrong though.
> > Maybe there could be a unit test which ensures that the flux reduces to the flat system for Minkowski?
> 
> Hi @nikwit - I found and fixed a major bug in the code here, the internal and external contributions were being added instead of subtracted. I wrote a test comparing this flux class with the flat `ScalarWave`'s upwind flux in flat space, and that helped me discover this bug. After fixing it, that test now passes. This code is now ready for review.

I also updated the executable in #2973  to use this upwind flux. We get stable evolutions of plane waves there now. Can I request you to have a quick look at this PR so we can get this in quick please?

**Comment by nilsvu** [comment]
@prayush would you mind deleting the old code you are porting over, if it's not needed anymore?

**Comment by prayush** [comment]
> @prayush would you mind deleting the old code you are porting over, if it's not needed anymore?

@nilsleiffischer : Removed the redundant code too, you're right we don't really need it it anywhere.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: is there a `not_null` version of this function so you can avoid memory allocations? Anything that does memory allocations on the interfaces is rather expensive. The `char_fields` function also does 2 more memory allocations, so overall this part of the boundary correction is already at 4 (including the shift dot normal below) memory allocations per interface plus a bunch of copy of data. It's very little work compared to memory movement.

(see comment below for more thoughts :) )

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp:201]: this function is also pretty brutal on memory allocations. In general, we should do at most 1 memory allocation per interface in `dg_boundary_terms` and one in `dg_package_data`. 

I'll let you and @nikwit coordinate what you think is best here since he's the one using the evolution system. If the conclusion is "let's merge this PR with all the extra allocations and copies" then it would be good to file an issue so that we remember this is a problem

File: tests/Unit/Evolution/Systems/CurvedScalarWave/CMakeLists.txt
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/CurvedScalarWave/CMakeLists.txt]: [optional] You can also change this to
```
  ""
  )

target_link_libraries(
  ${LIBRARY}
  PRIVATE
  # all the libs to link
  )
```
this is easier to read and what we're hoping to switch to in the long term :)

**Review by nilsdeppe — DISMISSED** [review]
Looks good, some thoughts that I think you and @nikwit should decide on the best course of action for. If you do end up changing code, please feel free to squash things directly :)

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (nikwit) [inline src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: There is a function `CurvedScalarWave::characteristic_speeds` which for some reason returns a `std::array<DataVector, 4>`. I would change it to the implementation you have right here and return a tensor, this could then also be called from the boundary conditions and makes `dg_package_data` very concise.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.hpp
Comment (nikwit) [inline src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.hpp:88]: `tnsr::a<3>` or `tnsr::i<4>`? Shouldnt make a difference as it is just used for packaging but I would personally prefer the latter.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (nikwit) [inline src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: you are relying on copy elision here but maybe a return by reference might be safer to avoid a huge amount of allocations if it fails.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/BoundaryCorrection.hpp
Comment (nikwit) [inline src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/BoundaryCorrection.hpp:1]: (random line) I think you need to add the tag `using inverse_spatial_metric_tag = gr::Tags::InverseSpatialMetric<volume_dim, Frame::Inertial, DataVector>;` to `System.hpp`, otherwise your normal vectors will be normalized w.r.t. the Euclidean metric resulting in the wrong computation of characteristic speeds, see `PackageDataImpl.hpp:48`. This will then also automatically slice in the normal vector in addition to the covector for reasons I dont understand. I suppose you might then not need the `inverse_spatial_metric` if you change the `characteristic_fields` function to accept the unit vector directly. Maybe @nilsdeppe can explain?

File: tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp
Comment (nikwit) [inline tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp]: I see no reason to have Tempbuffers in a unit test. They reduce readability and we do not care about allocations/performance, so I would argue to remove it.

**Review by nikwit — CHANGES_REQUESTED** [review]
I still find a discrepancy between what I implemented and this one. I will try and hunt down the difference this weekend, but here are some comments for now.

Update: I have found the discrepancy to be a bug in my comparison code. This implementation and the one I did, which is a copy of the `ScalarWave` system with only characteristic speed calculation changed, produce the same results for random values even in curved space.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.hpp
Comment (nilsdeppe) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.hpp:88]: This should be consistent with the other systems, which I think use a `tnsr::a<3>`

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/BoundaryCorrection.hpp
Comment (nilsdeppe) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/BoundaryCorrection.hpp:1]: if a system has a non-flat Cartesian metric then the `inverse_spatial_metric` tag must be specified in the system, otherwise you'll get incorrect normal vectors

Comment (prayush) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/BoundaryCorrection.hpp:1]: Just added the `inverse_spatial_metric_tag` to System, thanks!

File: tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp
Comment (prayush) [reply tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp]: Hmm. I sort of disagree on this. I would prefer leaving it here as is, if you don't mind too much. 

The reduction to readability of the test is limited to the declaration of the buffer itself, after which we are simply extracting references to memory in it - which just replaces allocating memory. 

And as spectre adds more and more unit tests, why would we not care about the performance of unit tests?

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (prayush) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: I would be happy to make this change but I worry this might not play well with other parts of the code, as all evolution systems at the moment use `std::array`'s. A quick grep tells me char speeds are directly used in, for e.g., the fluxes:

```
NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/LocalLaxFriedrichs.hpp
NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/Hll.hpp
```

@nilsdeppe can you please confirm that swapping out the type of characteristic speeds from `std::array<DataVector, 4> --> tnsr::a<3>` not break anything else?

File: tests/Unit/Evolution/Systems/CurvedScalarWave/CMakeLists.txt
Comment (prayush) [reply tests/Unit/Evolution/Systems/CurvedScalarWave/CMakeLists.txt]: changed

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (nilsdeppe) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: Those fluxes aren't used anymore by the evolution code (I don't think by the elliptic code either). However, since every other system's `characteristic_speeds` function uses `std::array`s it seems it would be better not to change it in just one system. It is possible to assign a `std::array` to a `tnsr::a<3>`, but it might have to be done explicitly: `*packaged_char_speeds = tnsr::a<3, DataVector, Frame::Inertial>{characteristic_speeds(...)};`

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (prayush) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: fair point, replaced with return by reference now

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (prayush) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: using return by reference of this function, AND removed the memory allocation in the `char_fields` function.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (prayush) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp:201]: Okay, reduced the total number of allocations to 1 buffer allocation. Also reduced the total size of allocation by reusing memory.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (prayush) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: I agree that this should change simultaneously in all evolution systems. And that is why I'd rather not call the `CurvedScalarWave::characteristic_speeds` function in `dg_package_data` as that would need an additional allocation of a `std::array<DV, 4>`. Its true this leads to code duplication, but I can't think of a better solution given the constraints.

**Comment by prayush** [comment]
Hi @nikwit @nilsdeppe : all your comments and suggestions have been incorporated / addressed. This PR is ready for another look.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (nikwit) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: ok, I would argue that avoiding this code duplication is more important than keeping a completely unused function consistent with other unused functions, which are now essentially deprecated. Keep in mind that I will have to duplicate it again for the boundary conditions as will pretty much any other boundary condition / boundary correction implementation.

But if you decide to leave it, I can port this over for all systems before I do the boundary conditions - I dont want to hold this PR up.

File: tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp
Comment (nikwit) [reply tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp]: I think since the build time of unit tests is far longer than the runtime and this allocation is not in any kind of loop, the reduced memory allocation speed-up will not be noticeable at all.

Overall I just feel strongly that unit tests should be as explicit and clear as possible making them easy to debug and useful for understanding how the tested code works. But I dont want to impose my opinions here too strongly so I would be fine with approving it like this.

**Comment by nikwit** [comment]
You have a python-related CI failure in gcc8 debug which is the one that uses python2, so there is probably some compatibility issue there.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (prayush) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: Okay, I am convinced. As a middle ground, I have added an overload of `characteristic_speeds` that returns a `tnsr::a<3>` by reference instead of a `std::array`. I have not deleted the version(s) that return(s) `array`s as that should be done in a different PR for all evolution systems simultaneously. 

This overload solves our purpose in both the flux and boundary conditions for this system.

File: tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp
Comment (prayush) [reply tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp]: That is fair. what also changed my mind is that this would be the only unit test that actually uses temp buffers, so I've removed their usage here as well.

**Comment by prayush** [comment]
Hi @nikwit : this is again ready for your perusal. @nilsdeppe already said above to squash all fixups directly. Let me know if you're okay with me squashing them now.

File: src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp
Comment (nikwit) [reply src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp]: I think that is a good solution

**Review by nikwit — DISMISSED** [review]
This looks good to me now, please go ahead and squash

**Comment by prayush** [comment]
@nikwit - just squashed all fixups.

**Review by nikwit — APPROVED** [review]
looks good. Out of curiosity (and because I know this will happen to me at some point), what was the reason for the python 2 failure?

**Comment by prayush** [comment]
> looks good. Out of curiosity (and because I know this will happen to me at some point), what was the reason for the python 2 failure?

Missing `__init__.py` files in the unit test directory!

**Comment by prayush** [comment]
Hi @sxs-collaboration/spectre-core-devs - this PR is ready for a look / merge.

**Review by kidder — APPROVED** [review]

**Pull request merged**
